velox/exec/NestedLoopJoinProbe.cpp

```diff
@@ -80,15 +113,221 @@ void NestedLoopJoinProbe::addInput(RowVectorPtr input) {
+            buildProjections_,
+            identityProjections_);
+        ++buildIndex_;
+      }
+      if (hasProbedAllBuildData()) {
+        setState(ProbeOperatorState::kFinish);
+      }
+      break;
+    }
+
+    if (input_ == nullptr) {
+      break;
+    }
+
+    // When input_ is not null but buildIndex_ is at the end, it means the
+    // matching of input_ and buildData_ has finished. For left/full joins,
+    // the next step is to emit output for mismatched probe side rows.
+    if (hasProbedAllBuildData()) {
+      output = needsProbeMismatch(joinType_) ? getMismatchedOutput(
+                                                   input_,
+                                                   probeMatched_,
+                                                   probeOutMapping_,
+                                                   identityProjections_,
+                                                   buildProjections_)
+                                             : nullptr;
+      finishProbeInput();
+      break;
+    }
+
+    const vector_size_t probeCnt = getNumProbeRows();
+    output = doMatch(probeCnt);
+    if (advanceProbeRows(probeCnt)) {
+      if (!needsProbeMismatch(joinType_)) {
+        finishProbeInput();
+      }
+    }
+  }
+  return output;
+}
+
+void NestedLoopJoinProbe::initializeFilter(
+    const core::TypedExprPtr& filter,
+    const RowTypePtr& probeType,
+    const RowTypePtr& buildType) {
+  VELOX_CHECK_NULL(joinCondition_);
+
+  std::vector<core::TypedExprPtr> filters = {filter};
+  joinCondition_ =
+      std::make_unique<ExprSet>(std::move(filters), operatorCtx_->execCtx());
+
+  column_index_t filterChannel = 0;
+  std::vector<std::string> names;
+  std::vector<TypePtr> types;
+  auto numFields = joinCondition_->expr(0)->distinctFields().size();
+  names.reserve(numFields);
+  types.reserve(numFields);
+  for (auto& field : joinCondition_->expr(0)->distinctFields()) {
+    const auto& name = field->field();
+    auto channel = probeType->getChildIdxIfExists(name);
+    if (channel.has_value()) {
+      auto channelValue = channel.value();
+      filterProbeProjections_.emplace_back(channelValue, filterChannel++);
+      names.emplace_back(probeType->nameOf(channelValue));
+      types.emplace_back(probeType->childAt(channelValue));
+      continue;
+    }
+    channel = buildType->getChildIdxIfExists(name);
+    if (channel.has_value()) {
+      auto channelValue = channel.value();
+      filterBuildProjections_.emplace_back(channelValue, filterChannel++);
+      names.emplace_back(buildType->nameOf(channelValue));
+      types.emplace_back(buildType->childAt(channelValue));
+      continue;
+    }
+    VELOX_FAIL(
+        "Join filter field {} not in probe or build input, filter: {}",
+        field->toString(),
+        filter->toString());
+  }
+
+  filterInputType_ = ROW(std::move(names), std::move(types));
+}
+
+RowVectorPtr NestedLoopJoinProbe::getMismatchedOutput(
+    RowVectorPtr data,
```

Comment:
use const &; please, go over the PR and make sure to pass shared pointers by reference.

Reply:
Sure, I've read through the changes and updated with the fix.